src/util/NuulsUploader.cpp

```diff
@@ -0,0 +1,223 @@
+        // Each entry in source->urls() is a QUrl pointing to a file that was copied.
+        for (const QUrl &path : source->urls())
+        {
+            QString localPath = path.toLocalFile();
+            if (!getImageFileFormat(localPath).isEmpty())
+            {
+                channel->addMessage(makeSystemMessage(
+                    QString("Uploading image: %1").arg(localPath)));
+                QImage img = QImage(localPath);
+                if (img.isNull())
+                {
+                    channel->addMessage(
+                        makeSystemMessage(QString("Couldn't load image :(")));
+                    return;
+                }
+
+                boost::optional<QByteArray> imageData = convertToPng(img);
+                if (imageData)
+                {
+                    RawImageData data = {imageData.get(), "png"};
+                    uploadQueue.push(data);
+                }
+                else
+                {
+                    channel->addMessage(makeSystemMessage(
+                        QString("Cannot upload file: %1, Couldn't convert "
+                                "image to png.")
+                            .arg(localPath)));
+                }
+            }
+            else if (localPath.endsWith(".gif"))
+            {
+                channel->addMessage(makeSystemMessage(
+                    QString("Uploading GIF: %1").arg(localPath)));
+                QFile file(localPath);
+                bool isOkay = file.open(QIODevice::ReadOnly);
+                if (!isOkay)
+                {
+                    channel->addMessage(
+                        makeSystemMessage(QString("Failed to open file. :(")));
+                    return;
+                }
+                RawImageData data = {file.readAll(), "gif"};
+                uploadQueue.push(data);
+                file.close();
+                // file.readAll() => might be a bit big but it /should/ work
+            }
+            else
+            {
+                channel->addMessage(makeSystemMessage(
+                    QString("Cannot upload file: %1, not an image")
+                        .arg(localPath)));
+                isUploading = false;
+            }
+        }
+        if (!uploadQueue.empty())
+        {
+            uploadImageToNuuls(uploadQueue.front(), channel, outputTextEdit);
+            uploadQueue.pop();
+        }
+    }
+    else
+    {  // not PNG, try loading it into QImage and save it to a PNG.
+        QImage image = qvariant_cast<QImage>(source->imageData());
+        boost::optional<QByteArray> imageData = convertToPng(image);
+        if (imageData)
+        {
+            uploadImageToNuuls({imageData.get(), "png"}, channel,
+                               outputTextEdit);
+        }
+        else
+        {
+            channel->addMessage(makeSystemMessage(
+                QString("Cannot upload file, failed to convert to png.")));
```

Comment:
I believe an `isUploading = false;` is missing in this block, right?

Reply:
yes

Reply:
yes